FEAT(client): Implement a mute cue

pkern opened the pull request:
The mute cue is played when speech is detected while the local client is
muted. This reminds the user that they are still muted, which is
especially useful when the overlay/client indication is not available
(e.g. in full-screen games).

This currently reuses the existing "audio off" cue.

Implements #4866


### Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

Krzmbrzl reviewed (changes requested):
I'll review the actual changes in the coming days but for now could you separate the translation changes into their own commit, please?

You can achieve that by using`git reset HEAD~1` then `git restore src/mumble/*.ts` and then commiting your code changes again.
After that, please run `scripts/updatetranslations.py` which will create a new commit for the translation changes. And finally push both of these commits to your branch (force-push).

Having the translation changes separate facilitates reviewing :)

pkern commented:
> I'll review the actual changes in the coming days but for now could you separate the translation changes into their own commit, please?
> Having the translation changes separate facilitates reviewing :)

Done. I mostly did it because the checks were yelling at me, but sure, I split them out. ;-)

Krzmbrzl commented on src/mumble/AudioInput.cpp:
```diff
@@ -943,6 +944,8 @@ int AudioInput::encodeCELTFrame(short *psSource, EncodingOutputBuffer &buffer) {
 	return len;
 }
 
+static const int MUTE_CUE_DELAY = 5000;  // in ms
```
I think I would turn this into a static member variable in the `AudioOutput` class. Also I would suggest to use `constexpr` instead of `const` for this kind of stuff.

Krzmbrzl replied:
Yeah sorry, I meant `AudioInput`

Krzmbrzl commented on src/mumble/AudioInput.cpp:
```diff
@@ -1072,8 +1075,12 @@ void AudioInput::encodeAudioFrame(AudioChunk chunk) {
+	bool bTalkingWhenMuted = false;
 	if (Global::get().s.bMute || ((Global::get().s.lmLoopMode != Settings::Local) && p && (p->bMute || p->bSuppress)) || Global::get().bPushToMute
 		|| (voiceTargetID < 0)) {
+		if (bIsSpeech) {
+			bTalkingWhenMuted = true;
+		}
```
This could be shortened as
```suggestion
		bTalkingWhenMuted = bIsSpeech;
```
The `if` makes the semantics a bit clearer though, so in the end I think I don't mind which version is used (I assume the extra branch will be optimized away anyways). Your call

pkern replied:
I just found what is arguably a bug: If we are deafened, the mute cue should not play. So I added that to the expression, PTAL.

Krzmbrzl commented on src/mumble/AudioInput.cpp:
```diff
@@ -6,6 +6,7 @@
 #include "AudioInput.h"
 
 #include "AudioOutput.h"
+#include "AudioOutputSample.h"
```
What is this needed for? :eyes:

pkern replied:
Removed. I had an incomplete type in an intermediate state where I tried to use slots.

Krzmbrzl commented on src/mumble/AudioInput.cpp:
```diff
@@ -1094,12 +1101,22 @@ void AudioInput::encodeAudioFrame(AudioChunk chunk) {
-		else if (ao && !bIsSpeech && bPreviousVoice)
-			ao->playSample(Global::get().s.qsTxAudioCueOff);
+
+		if (Global::get().s.bTxAudioCue) {
+			if (bIsSpeech && !bPreviousVoice && ao)
+				ao->playSample(Global::get().s.qsTxAudioCueOn);
+			else if (ao && !bIsSpeech && bPreviousVoice)
+				ao->playSample(Global::get().s.qsTxAudioCueOff);
+		}
+
+		if (Global::get().s.bTxMuteCue && ao && bTalkingWhenMuted) {
+			if (!qetLastMuteCue.isValid() || qetLastMuteCue.elapsed() > MUTE_CUE_DELAY) {
+				qetLastMuteCue.start();
+				ao->playSample(Global::get().s.qsTxMuteCue);
+			}
+		}
```
```suggestion
		if (ao) {
		    if (Global::get().s.bTxAudioCue) {
			      if (bIsSpeech && !bPreviousVoice) {
				      ao->playSample(Global::get().s.qsTxAudioCueOn);
			      } else if (!bIsSpeech && bPreviousVoice) {
				      ao->playSample(Global::get().s.qsTxAudioCueOff);
			      }
		    }
		    
		    if (Global::get().s.bTxMuteCue && bTalkingWhenMuted) {
			      if (!qetLastMuteCue.isValid() || qetLastMuteCue.elapsed() > MUTE_CUE_DELAY) {
				      qetLastMuteCue.start();
				      ao->playSample(Global::get().s.qsTxMuteCue);
			      }
		    }
		}
```

pkern replied:
Good point, done.

Krzmbrzl commented on src/mumble/AudioInput.h:
```diff
@@ -6,6 +6,7 @@
 #ifndef MUMBLE_MUMBLE_AUDIOINPUT_H_
 #define MUMBLE_MUMBLE_AUDIOINPUT_H_
 
+#include <QtCore/QElapsedTimer>
```
```suggestion
#include QElapsedTimer>
```
This is actually the preferred way of including Qt headers. We only found out about that very recently and therefore the rest of the code base partially does use the other style.

Krzmbrzl replied:
If clang-format is happy with it, we are too :)
Did you run clang-format yet? If not, pull the latest changes and then just run it on all files that you have edited. That should take care of all formatting aspects.

pkern replied:
Ran clang-format now. Nice stuff.

Krzmbrzl commented on src/mumble/Settings.cpp:
```diff
@@ -332,6 +334,9 @@ Settings::Settings() {
 	qsTxAudioCueOn  = cqsDefaultPushClickOn;
 	qsTxAudioCueOff = cqsDefaultPushClickOff;
 
+	bTxMuteCue  = false;
```
```suggestion
	bTxMuteCue  = true;
```
I think we can default to enabling this feature as it seems very handy. Therefore I think most folks like being warned about speaking when being muted and the ones that don't, can then go and turn it off for themselves.

pkern replied:
I did that, thanks. One case I could imagine where it's annoying is if you are passively listening (e.g. on the c3 lingo server). However in interactive cases it seems more useful. I also guess we don't have any infrastructure to easily toggle the default later, right? Then maybe it's better to start out this way and maybe offer a button in the top bar if that's something that needs toggling more often?

Krzmbrzl replied:
Wdym "changing it later"? We can change it in any release if we wish to do so.

> maybe offer a button in the top bar if that's something that needs toggling more often?

Yeah that sounds good

pkern replied:
I guess I sort of assumed that any original default would be written into the local user config. If we only persist differences, that'd be easy. ;-)

Do we need the button in this rev of the change or can we wait for beta feedback?

pkern reviewed (commented):
PTAL. I kept the review updates in a separate commit, let me know if I should merge it into the main one.

pkern commented:
I guess the commit-check is clear enough and this requires force-pushing of a squashed version.

pkern commented:
Rebased onto the format change. I also changed from `on.ogg` to `off.ogg`. I don't personally understand the value of the audio cue yet, but it probably makes sense when both are enabled to play the off sound rather than the "I'm now sending" cue.

Krzmbrzl commented on src/mumble/AudioInput.cpp:
```diff
@@ -1080,8 +1080,10 @@ void AudioInput::encodeAudioFrame(AudioChunk chunk) {
+	bool bTalkingWhenMuted = false;
 	if (Global::get().s.bMute || ((Global::get().s.lmLoopMode != Settings::Local) && p && (p->bMute || p->bSuppress))
 		|| Global::get().bPushToMute || (voiceTargetID < 0)) {
+		bTalkingWhenMuted = bIsSpeech && !(p && (p->bDeaf || p->bSelfDeaf));
```
1. What you want to check here is `Global::get().s.bDeaf` - that's where the local user's deaf status is (somewhat) reliably stored.
2. I think semantically the check for whether the user is deafened should not be done here as the user might still be talking while being muted. _But_ in that case we don't need to play the cue (though if we did there would probably be no difference since the user won't hear it anyways ^^). In short: I would check for this further down when you actually decide whether or not to play the cue.

pkern replied:
Ack, implemented both. Thank you!

Krzmbrzl commented on src/mumble/AudioInput.cpp:
```diff
@@ -1102,12 +1104,24 @@ void AudioInput::encodeAudioFrame(AudioChunk chunk) {
+
+		if (ao) {
+			if (Global::get().s.bTxAudioCue) {
+				if (bIsSpeech && !bPreviousVoice)
+					ao->playSample(Global::get().s.qsTxAudioCueOn);
+				else if (!bIsSpeech && bPreviousVoice)
+					ao->playSample(Global::get().s.qsTxAudioCueOff);
```
```suggestion
				if (bIsSpeech && !bPreviousVoice) {
					ao->playSample(Global::get().s.qsTxAudioCueOn);
				} else if (!bIsSpeech && bPreviousVoice) {
					ao->playSample(Global::get().s.qsTxAudioCueOff);
				}
```
When we're changing these already, we might as well add these curly brackets (we prefer always having them)

pkern replied:
Done.

Krzmbrzl commented on src/mumble/AudioInput.h:
```diff
@@ -227,6 +230,8 @@ class AudioInput : public QThread {
 	/// Number of 10ms audio "frames" per packet (!= frames in packet)
 	int iAudioFrames;
 
+	static constexpr unsigned int iMuteCueDelay = 5000;  // in ms
```
```suggestion
	/// The minimum time in ms that has to pass between the playback of two consecutive mute cues.
	static constexpr unsigned int iMuteCueDelay = 5000;
```

pkern replied:
Done, thanks!

Krzmbrzl commented:
Thank you very much for your contribution :+1:

lePoMo commented:
Is there a config for this ? I updated my mumble client (Mumble 1.4.230 released 2022-01) and cannot find where to disable this behavior.

Krzmbrzl commented:
Yes there is a setting under `Audio Input` (at the bottom right under the setting for `Audio cue`)

lePoMo commented:
Thank you. Not sure how I didn't see that. Also I guess was expecting it in "Messages", but I see how it is different to other things in messages.
